cpp/src/scalar/scalar_factories.cpp

```diff
@@ -62,4 +62,26 @@ std::unique_ptr<scalar> make_timestamp_scalar(
                                        stream, mr);
 }
 
+namespace {
+
+struct scalar_default_construction_helper {
```

Comment:
AFAICT the existence of this struct and the `make_default_constructed_scalar` is solely to produce the CUDF_FAIL for invalid types. Does this ever happen? Is it possible for a code path to hit this failure? e.g. are there CUDF data types that could be dispatched at run time that we can't create a scalar from?  If not, then I think the second `operator()` could be removed so that you get a compile-time error if it happens rather than a run-time error?